FactoryTools.class.nut

```diff
@@ -0,0 +1,156 @@
+    */
+    static function isFactoryFirmware() {
+        if ( _isAgent() )  {
+            // Running as factory agent
+            if (_isFactoryFirmware == null) {
+                // Agent doesn't have the information it needs
+                server.error("Agent use of Factory Tools requires the class to be also instantiated in device code");
+            }
+            return _isFactoryFirmware;
+        } else {
+            // Running on factory device
+            if ("factoryfirmware" in imp.configparams && imp.configparams["factoryfirmware"]) {
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+    * Factory Device: @return {bool} - 'true' or 'false'
+    * Factory Agent: @return {null | bool} - 'null' on error, or 'true' or 'false'
+    */
+    static function isFactoryImp() {
+        if ( _isAgent() )  {
+            // Running as factory agent
+            if (_isDeviceFixture == null) {
+                // Agent doesn't have the information it needs
+                server.error("Agent use of Factory Tools requires the class to be also instantiated in device code");
+            }
+            return _isDeviceFixture;
+        } else {
+            if (isFactoryFirmware()) {
+                // Running on a factory device
+                if ("factory_imp" in imp.configparams) {
+                    // Device is a BlinkUp fixture
+                    if (imp.configparams["factory_imp"] == imp.getmacaddress()) {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+    }
+
+
+    /**
+    * Factory Device: @return {bool} - 'true' or 'false'
+    * Factory Agent: @return {null | bool} - 'null' on error, or 'true' or 'false'
+    */
+    static function isDeviceUnderTest() {
+        if ( _isAgent() )  {
+            // Running as factory agent
+            if (_isProdDevice == null) {
+                // Agent doesn't have the information it needs
+                server.error("Agent use of Factory Tools requires the class to be also instantiated in device code");
+            }
+            return _isProdDevice;
+        } else {
+            // Running on a factory device
+            if (isFactoryFirmware() && !isFactoryImp()) {
+                // Device is a unit under test
+                return true;
+            }
+            return false;
+        }
+    }
+
+    /**
+    * Factory Device: @return {null}
+    * Factory Agent: @return {null | string} - if agent is Device under test the Factory Fixture agent URL is returned.
+    */
+    static function getFactoryFixtureURL() {
+        if("factory_fixture_url" in imp.configparams) {
+            return imp.configparams.factory_fixture_url;
+        }
+        return null;
+    }
+
+    /**
+    * @return {bool}
+    */
+    function _isAgent() {
+        // Running on an agent
+        if (imp.environment() == 2) return true;
```

Comment:
@smittytone was there a reason you were checking imp.environment with _2_ instead of _ENVIRONMENT_AGENT_?